CONTRIBUTING.md

```diff
@@ -0,0 +1,48 @@
+# How to contribute
+
+## Submitting PRs
+
+Please follow the provided PR template when writing a description for your changes.
+
+Do not take criticism personally. When you get feedback, it is about your work,
+not your character, personality, etc. Keep in mind we all want to make the project better.
+
+When something is not clear, please ask questions to clear things up.
+
+If you would like to introduce a big architectural changes or do a refactoring
+either in the codebase or the development tools, please open an issue with a proposal
+first. This way we can think together about the problem and probably come up
+with a better solution.
+
+## Coding conventions and guidelines
+
+### Commit messages
+
+* Always write descriptive commit messages ("fix bug" is not acceptable).
+* Use the present tense ("Add feature" not "Added feature").
+* Use the imperative mood ("Move cursor to..." not "Moves cursor to...").
+* Limit the first line to 72 characters or less.
+* Include the number of the issue you are fixing.
+
+### Coding guidelines
+
+As a Python project, we must follow PEP8 guidelines.
+
+Furthermore, follow the Clean code conventions. The most important
+in this project are the following rules:
+
+* Simpler is always better.
```

Comment:
Near on the top to that list, I would add ...

- document the objects you implement (function, classes ...) using python doc-strings [PEP-0257](https://www.python.org/dev/peps/pep-0257/)

Reply:
I would not add comments to functions and classes, only to explain hard to understand stuff or warnings about a usage of a function/class. Clean code states that your code should be self-documenting. Also, comments are additional lines which you have to maintain with your code.

If someone really wants to comment their code, I do not want to forbid it. But in general I consider it bad practice especially in case of projects like searx. I understand that e.g. REST APIs require comments/documentation. But searx is not like that. :)

I would not add it to the list and I would not require anyone to comment their code.

Reply:
Wow, that's all sounds strange to me .. 

>  Clean code states that your code should be self-documenting.

If you have a bug in your function, the code self-documents the bug .. IMO: a function ensures treatment and that treatment should be described. / there is no need to document "c +=1" is the incrementation of "c" by one.

> Also, comments are additional lines which you have to maintain with your code.

The number of lines of code should not a criteria / My irst imperative is that the code must be legible.

FWIW: Do not mix comments and docstrings these are different subjects.

> If someone really wants to comment their code, I do not want to forbid it. But in general I consider it bad practice especially in case of projects like searx

It is bad practice to document code?  NO

> I would not add it to the list and I would not require anyone to comment their code.

Do not think in absolute categories, there is a lot of room between not to document at all and to document central functions.  Lets encourage user to document their code .. BTW: does anyone have a clear view on the language mess / shouldn't all the implicit names in plugins and engines not be documented?

Reply:
By example: 

PR https://github.com/asciimoo/searx/pull/1938 implements a layer for _resources_ which is documented,. If you want to see what _files and URLs_ a theme does manage, you can read its class documentation: https://return42.github.io/searx/dev/sources.html#searx.resources.Theme / there you can see e.g. what a `favicons_folder` is, other examples are `result_templates` or the `static_url`.

How should a new contributor know all this without a word of documentation? In absence of any documentation and the very imperative implementation I had to do a deep analysis first, this costs numerous days and also has the potential to fail.

Reply:
I agree that documentation would be a good thing at least on the core components. There is no need for a huge amount of documentation, but here and there it may help.

For example:
* search_query is [SearchQuery object](https://github.com/asciimoo/searx/blob/bce3830b8dd9322e4e93d0003aa86e03c03e2dca/searx/query.py#L177):
https://github.com/asciimoo/searx/blob/bce3830b8dd9322e4e93d0003aa86e03c03e2dca/searx/search.py#L414 
* search_query contains a query attribute:
https://github.com/asciimoo/searx/blob/bce3830b8dd9322e4e93d0003aa86e03c03e2dca/searx/query.py#L182

It is not clear what ```query``` is. It's only after looking at search.py that I remember what it is:
https://github.com/asciimoo/searx/blob/bce3830b8dd9322e4e93d0003aa86e03c03e2dca/searx/search.py#L271
and 
https://github.com/asciimoo/searx/blob/bce3830b8dd9322e4e93d0003aa86e03c03e2dca/searx/query.py#L166-L170

(I know that how ```query_parts``` is built).

Some docstring would help everyone here (or at least me). Perhaps also some typing.

Reply:
> Wow, that's all sounds strange to me ..

>>    Clean code states that your code should be self-documenting.

I recommend you to read [Uncle Bob's Clean Code](https://www.amazon.com/Clean-Code-Handbook-Software-Craftsmanship/dp/0132350882) . He explains the rationale behind this statement.

> If you have a bug in your function, the code self-documents the bug ..

Exactly! And I think it is a good thing. It increases the possibility of finding a bug.
There are numerous articles in this topic, [this one](https://blog.usejournal.com/stop-writing-code-comments-28fef5272752) summarizes the issues quite well IMHO.
Luckily python is a very readable programming language and it allows us to write easily understandable code which can document itself. I would rather emphasize this in the contrib guideline/PR template than making docstrings mandatory.

> How should a new contributor know all this without a word of documentation?

As I understand, @kvch doesn't say that there is no room for any kind of documentation, just that it isn't have to be mandatory in all cases - and I agree. There are cases where a small docstring can help but it doesn't mean that it should be mandatory. Also, we can ask for comments/clarifications any time during code reviews in specific cases.

@dalf your examples are good demonstration of ambiguous code parts, but in these cases comments/docstrings are not the only solution (neither the best in my opinion). More verbose and precise variable names can also help to understand the code. That's one of the things why we do code reviews - to detect and eliminate misunderstandable code.

Reply:
@asciimoo I will not try to disprove Uncle Bob. Yes first the code most be clear.

( If I continue the example about SearchQuery.query: 
* ```RawTextQuery.getSearchQuery()``` is misleading because it returns ```bytes``` not a ```SearchQuery``` object. 
* ```query``` is actually the user query without any modifiers (bang, language, timeout). So it should be renamed ```query_without_modifiers``` (there is for sure a better idea). So ```RawTextQuery.getSearchQuery()``` should be named ```RawTextQuery.getQueryWithoutModifiers()```.
* What about the encoding of SearchQuery.query, bytes or str ? 
* and, probably ```RawTextQuery``` should be name ```UserQueryParser``` and perhaps (?) move this code next to ```get_search_query_from_webapp``` in dedicated module about user input parsing, and .... and....

This is something I would like to do.

)

* My wish is to provide a clean API for the engines : some documentation can helpful in my opinion (related to https://github.com/asciimoo/searx/issues/1802 for example)

* Question: is typing documentation or code in Python ? 

* Another example searx/utils.py which provides functions / tools for webapp.py and the engines:
https://github.com/asciimoo/searx/blob/85b37233458c21b775bf98568c0a5c9260aa14fe/searx/utils.py#L361 
since the code is optimized (time critical function at the searx start), most probably it's where some comments would be helpful ?

Reply:
> As I understand, @kvch doesn't say that there is no room for any kind of documentation, just that it isn't have to be mandatory in all cases - and I agree. 

We are more or less in the same line / This is more or less what I mean ... may I mislead @kvch sentence "But in general I consider it bad practice especially in case of projects like searx."

To make it more exemplary for me: I linked an example, when it (IMHO) make sense to document names  ..  IMHO it make sense to document all the globals and implicit names, we have on objects like engines, plugins, themes .. . do you think different?

Reply:
>  since the code is optimized (time critical function at the searx start), most probably it's where some comments would be helpful ?

I agree, sometimes it's easier/more effective to write a comment than code 5 other functions just to make the code more understandable. What I wanted to say is that I agree with @kvch that writing comments shouldn't be a **required** thing in every single function/PR. I also agree with you that in some cases comments can be useful.
Side note: I don't think neither that the words of Uncle Bob are sacred and unquestionable, but most of the time it can be a good baseline to write maintainable and understandable code.
I don't have problem with adding a section to contrib guide about occasional comments, but I'd rather encourage contributors to try to write self-documenting code if it is possible.

Reply:
>  I don't think neither that the words of Uncle Bob are sacred and unquestionable

I've edited my message too much and removed the smiley trying to saying "don't take the sentence before too seriously".

Reply:
I agree with your comments / BTW: Uncle Bob names "good" and "bad" comments in [Chapter 4](https://www.investigatii.md/uploads/resurse/Clean_Code.pdf)